src/ast/InheritanceForest.java

```diff
@@ -75,21 +83,64 @@ public InheritanceForest(Program prog) {
 		}
 		
 		LLclasses.node curr;
+		String superName;
+		//meant for semantic check
+		int nodeMapPrevSize = nodeMap.size();
+
 		/*add all other classes to the forest*/
 		while(classBag.head != null) {
-			if(nodeMap.containsKey(classBag.head.value.superName())){
+			superName = classBag.head.value.superName();
+			//semantic check - can't extend main class
+			if(superName.equals(mainClass.name())){
+				isLegalTree = false;
+				return;
+			}
+			if(nodeMap.containsKey(superName)){
+				//semantic check - super is defined before child
+				if(nodeMap.get(superName).value.lineNumber >= classBag.head.value.lineNumber){
+					isLegalTree = false;
+					return;
+				}
+				//semantic check - no name repetition
+				if(nodeMap.containsKey(classBag.head.value.name())){
+					isLegalTree = false;
+					return;
+				}
+
 				addToForest(classBag.head.value);
 				classBag.removeHead();
 				continue;
 			}
 			curr = classBag.head;
 			while(curr.next != null) {
-				if(nodeMap.containsKey(curr.next.value.superName())) {
+				superName = curr.next.value.superName();
+				if(nodeMap.containsKey(superName)) {
+					//semantic check - super is defined before child
+					if(nodeMap.get(superName).value.lineNumber >= curr.next.value.lineNumber){
+						isLegalTree = false;
+						return;
+					}
+					//semantic check - no name repetition
+					if(nodeMap.containsKey(curr.next.value.name())){
+						isLegalTree = false;
+						return;
+					}
+
 					addToForest(curr.next.value);
 					classBag.removeNext(curr);
 				}
 				else curr = curr.next;
 			}
+
+			/*semantic check
+			if no classes were added (we know that there are classes to add because we entered the while)
+			than this means that there is at least one class with undefined super
+			 */
```

Comment:
I would check this explicitly. I think we should collect the names of *all* class names into a hash set and check against it whenever some class claims to extend another. Implicit checks like this although documented are very bug prone.

Reply:
this will increase the runtime bot I agree with the possible bugs so I'll take care of it.